src/middlewared/middlewared/plugins/smb.py

```diff
@@ -215,9 +215,19 @@ class Config:
     @private
     async def smb_extend(self, smb):
         """Extend smb for netbios."""
-        smb['netbiosname_local'] = smb['netbiosname']
-        if not await self.middleware.call('system.is_freenas') and await self.middleware.call('failover.node') == 'B':
-            smb['netbiosname_local'] = smb['netbiosname_b']
+
+        ha_mode = SMBHAMODE[(await self.get_smb_ha_mode())]
+
+        if ha_mode == SMBHAMODE.STANDALONE:
+            smb['netbiosname_local'] = smb['netbiosname']
+
+        elif ha_mode == SMBHAMODE.LEGACY:
+            failover_node = await self.middleware.call('failover.node')
+            smb['netbiosname_local'] = smb['netbiosname'] if failover_node == 'A' else smb['netbiosname_b']
+
+        elif ha_mode == SMBHAMODE.UNIFIED:
+            ngc = await self.middleware.call('network.configuration.config')
+            smb['netbiosname_local'] = ngc['hostname_virtual']
```

Comment:
Do we validate that `hostname_virtual` is provided at this point? I can't remember.

Reply:
This is the extend method so validation doesn't happen here.

Reply:
Derp